core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -0,0 +1,289 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.joiner;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.JoinConfig;
+import co.cask.cdap.etl.api.JoinElement;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.StageConfigurer;
+import co.cask.cdap.etl.api.batch.BatchJoiner;
+import co.cask.cdap.etl.api.batch.BatchJoinerContext;
+import co.cask.cdap.etl.api.batch.BatchJoinerRuntimeContext;
+import co.cask.hydrator.plugin.common.KeyValueListParser;
+import com.google.common.base.Splitter;
+import com.google.common.base.Strings;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Table;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.Path;
+
+/**
+ * Batch joiner to join records from multiple inputs
+ */
+@Plugin(type = BatchJoiner.PLUGIN_TYPE)
+@Name("Joiner")
+@Description("Performs join operation on records from each input based on required inputs. If all the inputs are " +
+  "required inputs, inner join will be performed. Otherwise inner join will be performed on required inputs and non " +
+  "records from non-required inputs will only be present if they match join criteria.")
+public class Joiner extends BatchJoiner<StructuredRecord, StructuredRecord, StructuredRecord> {
+  private final JoinerConfig conf;
+  private Map<String, Schema> inputSchemas;
+  private Schema outputSchema;
+  private Map<String, List<String>> perStageJoinKeys;
+  private Table<String, String, String> perStageFieldsToRename;
+  private Iterable<String> requiredInputs;
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_MAP = new TypeToken<Map<String, Map<String, String>>>() {
+  }.getType();
+
+  private static final Gson GSON1 = new GsonBuilder()
```

Comment:
unused?
Does the used GSON object not need custom type adapters? For Schema? Structured Record?